docs/source/advanced.rst

```diff
@@ -178,26 +178,126 @@ This is basically the same as calling :py:func:`~scikeras.wrappers.BaseWrapper.g
+
+.. code-block::
+
+                                   ↗ feature_encoder ↘
+    SciKeras.fit(features, labels)                    dataset_transformer → keras.Model.fit(data)
+                                   ↘ target_encoder  ↗ 
+
+
+Within SciKeras, this is roughly implemented as follows:
+
+.. code:: python
+
+    from sklearn.preprocessing import FunctionTransformer
+
+    from scikeras.utils.transformers import (
+        ClassifierLabelEncoder,
+        RegressorTargetEncoder,
+        ClassWeightDataTransformer
+    )
+
+
+    class BaseWrapper:
+
+        @property
+        def target_encoder(self):
+            return ClassifierLabelEncoder(loss=self.loss)
+        
+        @property
+        def feature_encoder(self):
+            return FunctionTransformer()
+        
+        @property
+        def dataset_transformer(self):
+            return FunctionTransformer()
+
+        def fit(self, X, y, sample_weight):
+            self.target_encoder_ = self.target_encoder
+            self.dataset_transformer_ = self.feature_encoder
+            self.feature_encoder_ = self.feature_encoder
+            y = self.target_encoder_.fit_transform(y)
+            X = self.feature_encoder_.fit_transform(X)
+            X, y, sample_weight = self.dataset_transformer_.fit_transform((X, y, sample_weight))
+            self.model_.fit(x=X, y=y, sample_weight=sample_weight)  # tf.keras.Model.fit
+            return self
+
+        def predict(self, X):
+            X = self.feature_encoder_.transform(X)
+            X, _, _ = self.dataset_transformer_.fit_transform((X, None, None))
+            y_pred = self.model_.predict(X)
+            return self.target_encoder_.inverse_transform(y_pred)
+
+    class KerasClassifier(BaseWrapper):
+
+        @property
+        def target_encoder(self):
+            return ClassifierLabelEncoder(loss=self.loss)
+        
+        @property
+        def dataset_transformer(self):
+            return ClassWeightDataTransformer(class_weight=self.class_weight)
+        
+        def predict_proba(self, X):
+            X = self.feature_encoder_.transform(X)
+            X, _, _ = self.dataset_transformer_.fit_transform((X, None, None))
+            y_pred = self.model_.predict(X)
+            return self.target_encoder_.inverse_transform(y_pred, return_proba=True)
+
+
+    class KerasRegressor(BaseWrapper):
+
+        @property
+        def target_encoder(self):
+            return RegressorTargetEncoder()
+
+
+One important thing to note is that ``feature_encoder`` and ``target_encoder``
+are run before building the Keras Model, while ``data_transformer`` is run after
+the Model is built. This means that the former two will not have access to the Model
+(eg. to get the number of outputs) but *will* be able to inject data into the model building
+function (more on this below). On the other hand,
+``data_transformer`` *will* get access to the built Model, but it cannot pass any data to model building
+function.
```

Comment:
> One important thing to note is that feature_encoder and target_encoder are run before building the Keras Model, while data_transformer is run after the Model is built.

This isn't clear in the example (though it's clear in the edit I've made).

> This means that the former two will not have access to the Model
(eg. to get the number of outputs) but *will* be able to inject data into the model building
function (more on this below). On the other hand,
``data_transformer`` *will* get access to the built Model, but it cannot pass any data to model building
function.

I would only say "the output of `dataset_transformer` get directly passed to tf.keras.Model.fit through self.model_.fit." All the about the number of outputs/etc is a confusing. I think I'd refer directly to the advanced examples.

Reply:
> This isn't clear in the example (though it's clear in the edit I've made).

Thank you for pointing that out! The example is much better with this detail.

> I would only say

That seems reasonable, I'll cut it down to ~ what you are suggesting.

Reply:
Thanks. Nothing I say hard and fast, all my comments are comments are illustrations. If a change to be hard and fast, I'll submit a PR.